### ISSUE TITLE
Effect typing of session exception handlers

### ISSUE THREAD

**Issue by SimonJF** [links-lang/links#1078]
Documenting some recent discussion with @slindley but happy to have input. 

Session exceptions are a restricted class of exception handler which correctly propagate failure along (linear) session endpoints, allowing us to maintain session progress in the presence of exceptions [see Fowler et al, POPL 2019].

Session exceptions are implemented through an elaboration pass to effect handlers. In doing so, we have a `SessionFail` effect which is introduced by `raise`, and `try-as-in-otherwise` is elaborated into an effect handler which handles `SessionFail` and cancels the session endpoints in its continuation.

The core calculus uses unchecked exceptions, but because of the implementation strategy in Links, we need `SessionFail` to appear in types explicitly. In theory, having session exceptions manifest in the types is the right decision.

However, there are a few issues:

  * Independently of anything else, `receive` is typed incorrectly -- a failing receive will raise an exception, so `receive` should have the `SessionFail` type when `session_exceptions` is enabled
  * Types change based on whether `session_exceptions` is enabled, regardless of whether session exceptions are actually used, since a top-level handler is installed on each process. This means certain type signatures need to be made more general than we might want.
  * Just more generally, a manifest `SessionFail` effect is very unpleasant to program with -- what should be quite simple programs to write now need the developer to be thinking about effect signatures on their types. Anecdotally I've found myself fighting far more with the effect system since having to worry about the effect

The complexity of the manifest `SessionFail` effect means that the Linear MVU patch, for example, still isn't in the main branch.

One particular proposal @slindley and I have is simply hiding this implementation mechanism in the `wild` effect. I'd be interested in any thoughts.

**Comment by dhil** [comment]
If `SessionFail` is meant to be an intrinsic effect, then it needs to have a name that cannot be typed by the programmer in Links source code, e.g. `_SessionFail`, otherwise you can accidentally capture the operation with a user-definable handler, which I suppose can mess up the entire session exceptions infrastructure.

**Comment by SimonJF** [comment]
Good point, but I believe this is sorted already: https://github.com/links-lang/links/blob/master/core/typeSugar.ml#L3999

**Comment by SimonJF** [comment]
That said -- I think your point more pertains to the actual handlers themselves -- and you're probably right that it is probably possible for a user to define a handler which can handle `SessionFail`. So the simple fix of making it `_SessionFail` is likely worth it anyway. Thanks!

**Comment by dhil** [comment]
Regarding what-to-do about the manifesting operation. In the short term we can certainly embed it in `wild`, it would be hack and it would introduce some special cases in the code base. Nevertheless doable, I reckon. However, I think in the long term we would want something more principled. The problem you are encountering is not unique, it happens time and time again when you naturally grow a code base. Essentially, the problem is that you are changing the signature of a bunch of core functions, and thus breaking legacy uses of those function. There is a research question to be explored here, namely, how to retroactively introduce effects?

Relatedly, I would like to eventually integrate all of the session exceptions into the handler stuff, meaning I would like to be able to pattern match on a session exception in my handlers. In addition, I would like to have some general notion of cancellation, i.e. the ability to abort an arbitrary continuation.
